### PR TITLE
Disable tool method

### PR DESCRIPTION
**Pull request by fivegrant**
This PR adds a new `disable` method to the ReAct agent. `disable` takes a string, matches the name to a tool and disables it.

## Advantages
There are many different ways to accomplish tool toggling via
- a special Beaker tool function `toggleable_tool`.
- a new option `enabled` to `tool`.
- a list `disabled_tools` passed into agent initialization.

However, the `disable` has various advantages over the previous approaches:
- the feature itself has a very small footprint in terms of both LoC and actual architectural changes. 
- the agent can have its tools disabled _after_ initialization. `disable` is much more flexible.
- `disable` accepts multiple positional arguments, therefore, it can be used identically to the `disabled_tools` approach. So instead of having `super.__init__(..., disabled_tools = ["some_tool", "another_tool"] ...)`, you can explicitly put `self.disables("some_tool", "another_tool")`. I'd argue the latter is actually a bit cleaner.
- (bonus) implementing a 'reenable/toggle' for tools is very trivial

TL;DR: The `disable` method is the simplest approach without sacrificing flexibility

### CODE REVIEW

File: archytas/react.py [hunk 1]
```diff
@@ -118,6 +119,23 @@ def __init__(
...
+                pattern = re.compile(f"\S+\.{tool_name}$")
+                matches = list(filter(lambda name: pattern.match(name), self.tools.keys()))
+                if len(matches) > 1:
+                    raise ValueError(f"Ambiguous name: Multiple tools called '{tool_name}'")
+                elif len(matches) == 1:
+                    tool_name = matches[0]
+            if tool_name in self.tools:
+                self.tools.pop(tool_name)
+            else:
+                raise AttributeError(f"Cannot disable nonexistent tool: {tool_name}")
```
Comment (mattprintz) [inline archytas/react.py:134]: This error seems less than ideal. The intention is to make sure that the tool isn't enabled. If it's not there, it's not enabled. Whether it has been already removed or was disabled somewhere else, or just never existed shouldn't matter.
Correct behavior is probably to just add a default to pop and let it not change self.tools if there's nothing to remove.

File: archytas/react.py [hunk 2]
```diff
@@ -118,6 +119,23 @@ def __init__(
...
+        self.prompt = build_prompt(self.tools)
+        self.system_message["content"] = self.prompt
```
Comment (mattprintz) [inline archytas/react.py:136]: These two lines seem to be too low-level for this function. It feels like there should be something like `update_prompt()` that doesn't need to take self.tools because that's already on self, and is also called as part of the setup process after the tools are defined.

File: archytas/react.py [hunk 3]
```diff
@@ -118,6 +119,23 @@ def __init__(
         # keep track of the last tool used (for error messages)
         self.last_tool_name = ""
 
+    def disable(self, *tool_names):
+        for tool_name in tool_names:
+            if "." not in tool_name:
+                pattern = re.compile(f"\S+\.{tool_name}$")
+                matches = list(filter(lambda name: pattern.match(name), self.tools.keys()))
```
Comment (mattprintz) [inline archytas/react.py:126]: Do we really need a regex here? How is the regex version substantively different than
`matches = [name for name in self.tools.keys() if name.endswith(f".{tool_name")]`?

File: archytas/react.py [hunk 4]
```diff
@@ -118,6 +119,23 @@ def __init__(
         # keep track of the last tool used (for error messages)
         self.last_tool_name = ""
 
+    def disable(self, *tool_names):
+        for tool_name in tool_names:
+            if "." not in tool_name:
+                pattern = re.compile(f"\S+\.{tool_name}$")
```
[the remark this answers is not in the document]
Comment (fivegrant) [reply archytas/react.py:125]: The case where it could fall through is if there are two tools "foo" and "SomeClass.foo". Otherwise, `tool_name` isn't overwritten. [The logic is a bit weird here though](https://github.com/jataware/archytas/blob/98e4ed87cc0b08a9b506151f1966f4117811f008/archytas/react.py#L127-L130). Note that we do nothing when no match is found and we keep the old tool name but that's far from obvious. I'll work on cleaning up this section

File: archytas/prompt.py [hunk 5]
```diff
@@ -57,7 +57,7 @@ def notes(*, ask_user: bool) -> str:
 """.strip()
 
 
-def build_prompt(tools: list[Callable]) -> str:
+def build_prompt(tools: list[Callable] | dict[str, Callable]) -> str:
```
[the remark this answers is not in the document]
Comment (fivegrant) [reply archytas/prompt.py:60]: The issue is kinda subtle. So, [the tools that are passed into build prompt](https://github.com/jataware/archytas/blob/36c8ace7400b0b81abb24e7fa77a75289b6b0c5e/archytas/react.py#L107) are different from  [self.tools](https://github.com/jataware/archytas/blob/36c8ace7400b0b81abb24e7fa77a75289b6b0c5e/archytas/react.py#L93). `build_prompt` uses the tools list passed in during initialization. The self.tools dictionary is generated from this list. IDK how much of this was clear before so my apologies if I'm explaining something obvious.

Anyway, I should probably have left `build_prompt` untouched and should have made the conversion in `self.disable` itself. When I implement `ReAct.update_prompt`, I'll revert `archytas/prompt.py`

File: archytas/react.py (hunk 2 above)
Comment (fivegrant) [reply archytas/react.py:136]: Good point. I'll implement `ReAct.update_prompt`

File: archytas/react.py (hunk 3 above)
Comment (mattprintz) [reply archytas/react.py:126]: @fivegrant Oh, I missed the first time around, but you need to be really careful with (or really just never do) string interpolation on raw regular expression strings.
Not only is it really easy for passed in data to cause an exception, it can crash your entire system.
Also, the string should be a raw string, and apparently raw fstrings are a thing, but yeah, still shouldn't try to interpolate in the thing you're trying to match on unless it's heavily sanitized.

File: archytas/react.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fivegrant) [reply archytas/react.py]: I tried to make the logic clearer with the recent push

**Review by mattprintz — APPROVED** [review]
:+1:
